core/templates/pages/contributor-dashboard-page/modal-templates/translation-modal.controller.ts

```diff
@@ -100,6 +146,93 @@ angular.module('oppia').controller('TranslationModalController', [
       $scope.activeWrittenTranslation.html = '';
     };
 
+    $scope.getElementTexts = function(elements, type) {
+      const textWrapperLength = 6;
+      const attributes = Array.from(elements, function(element: HTMLElement) {
+        if (element.localName === 'oppia-noninteractive-image') {
+          const attribute = element.attributes[type].value;
+          return attribute.substring(
+            textWrapperLength, attribute.length - textWrapperLength);
+        }
+      });
+      return attributes.filter(attribute => attribute);
+    };
+
+    $scope.notCopiedAllElements = function(
+        originalElements, translatedElements) {
+      const hasMatchingTranslatedElement = (element) => (
+        translatedElements.includes(element));
+      return !originalElements.every(hasMatchingTranslatedElement);
+    };
+
+    $scope.changedImgDetails = function(originalElements, translatedElements) {
+      if (originalElements.length === 0) {
+        return false;
+      }
+      const hasMatchingTranslatedElement = (element) => (
+        translatedElements.includes(element) && originalElements.length > 0);
+      return originalElements.some(hasMatchingTranslatedElement);
+    };
+
+    $scope.getTexts = function(rawText) {
+      const foundImageFilePaths = $scope.getElementTexts(
+        rawText, 'filepath-with-value');
+      const foundImageAlts = $scope.getElementTexts(
+        rawText, 'alt-with-value');
+      const foundImageDescriptions = $scope.getElementTexts(
+        rawText, 'caption-with-value');
+
+      return {
+        foundImageFilePaths,
+        foundImageAlts,
+        foundImageDescriptions
+      };
+    };
+
+    $scope.isTranslationCompleted = function(
+        originalElements: [HTMLElement], translatedElements: [HTMLElement]) {
+      const originalHtmlElements = Array.from(
+        originalElements, element => element.nodeName);
+      const translatedHtmlElements = Array.from(
+        translatedElements, element => element.nodeName);
+
+      for (const originalElement of new Set(originalHtmlElements)) {
+        if (
+          !(translatedHtmlElements.some(
+            element => element === originalElement)) ||
+          originalHtmlElements.filter(
+            element => element === originalElement).length >
+            translatedHtmlElements.filter(
+              element => element === originalElement).length
+        ) {
+          return false;
+        }
+      }
```

Comment:
This seems really convoluted. Is the array of elements guaranteed to be in sorted in the order in which they appear? If so, we can just compare the tag one by one?

Reply:
I removed `new Set(originalHtmlElements)` and used `originalHtmlElements `. I think now it makes sense. Doesn't it?

What this code snippet does: 
- First it checks whether an element of the `originalHtmlElements ` does not exist in the `translatedHtmlElements `.
- Then it checks whether the number of time a particular element is repeated in `originalHtmlElements ` is greater than it is in `translatedHtmlElements`.
- If any of the above criteria is matched, it returns as the translation is not completed.